packages/venia-ui/lib/components/Extension/extension.js

```diff
@@ -0,0 +1,30 @@
+import React, { Fragment, useEffect, useState } from 'react';
+import { createPortal } from 'react-dom';
+
+/**
+ * This extension component renders children into the DOM after React mounts.
+ *
+ * @example
+ * <Extension targetId={'extension-point-1'}>
+ *   <MyExtensionComponent>
+ * </Extension>
+ *
+ * @param {String} [props.targetId=root] id of injection target for extension
+ * @param {ReactNode} props.children
+ */
+export const Extension = props => {
+    const { children, targetId = 'root' } = props;
+    const [isMounted, setMounted] = useState(false);
+
+    useEffect(() => {
+        setMounted(true);
+    }, []);
+
+    const extensionPoint = document.getElementById(targetId);
+
+    // Only render after the initial mount, and only if the extension point has
+    // been found.
+    return isMounted && !!extensionPoint
+        ? createPortal(<Fragment>{children}</Fragment>, extensionPoint)
+        : null;
+};
```

Comment:
Reading up on Portals it seems there is an alternative method to mounting which doesn't require the target to have been mounted. The following does still wait to append/remove the extension DOM until mount but it immediately creates the DOM node on which to stick the `children`. I _think_ we should stick with what we have as it's probably a better experience for extension developers who may want to access state/size/etc but I just wanted to point out an alternative render approach.

```js
export const Extension = props => {
    const { children, targetId = 'root' } = props;
    const extensionRoot = useMemo(() => document.createElement('div'), []);

    useEffect(() => {
        const extensionPoint = document.getElementById(targetId);
        extensionPoint.appendChild(extensionRoot);
        return () => {
            extensionPoint.removeChild(extensionRoot);
        }
    }, [extensionRoot, targetId]);

    return createPortal(children, extensionRoot);
};
```